### PR TITLE
Adding webhook for readiness provider

### PR DESCRIPTION
**Pull request by sathyanarays**
### What this PR does / why we need it
Apart from the OpenAPI spec validations, additional checks need to be performed to validate readiness provider resources. This PR adds the validating webhook to both main function and the envtest

### Which issue(s) this PR fixes

Fixes #

### Describe testing done for PR
Manual testing and envtests

### Release note

```release-note

```

### CODE REVIEW

File: apis/core/v1alpha2/readinessprovider_webhook.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+var _ webhook.Validator = &ReadinessProvider{}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type
+func (r *ReadinessProvider) ValidateCreate() error {
```
Comment (avi-08) [inline apis/core/v1alpha2/readinessprovider_webhook.go:40]: Another validation for create and update operation is whether names in `checkRefs` exist in any readiness resource definition

Comment (sathyanarays) [reply apis/core/v1alpha2/readinessprovider_webhook.go:40]: Readiness Provider can exist without a corresponding readiness definition.

File: packages/readiness/bundle/config/upstream/rbac.yaml [hunk 2]
```diff
@@ -69,6 +69,24 @@ rules:
...
+    verbs:
+      - get
+      - list
+      - patch
+      - update
+  - apiGroups:
+      - admissionregistration.k8s.io
+    resources:
+      - validatingwebhookconfigurations
+      - mutatingwebhookconfigurations
```
[the remark this answers is not in the document]
Comment (sathyanarays) [reply packages/readiness/bundle/config/upstream/rbac.yaml:85]: This is required as the controller tries to list both validating and mutating webhooks!

File: apis/core/v1alpha2/readinessprovider_webhook.go [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	if oldObj == nil {
+		return nil
+	}
+
+	return nil
+}
+
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *ReadinessProvider) ValidateDelete() error {
```
[the remark this answers is not in the document]
Comment (sathyanarays) [reply apis/core/v1alpha2/readinessprovider_webhook.go:68]: I feel there is nothing to validate while deletion. Please suggest if there are any validations need to be done!

File: apis/core/v1alpha2/readinessprovider_webhook.go (hunk 1 above)
Comment (rajathagasthya) [reply apis/core/v1alpha2/readinessprovider_webhook.go:40]: The check references need to exist though, right? You can't give invalid names there.

File: apis/core/v1alpha2/readinessprovider_webhook.go [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *ReadinessProvider) ValidateUpdate(old runtime.Object) error {
+	readinessproviderlog.Info("validate update", "name", r.Name)
+
+	oldObj, ok := old.(*ReadinessProvider)
+	if !ok {
+		return apierrors.NewBadRequest(fmt.Sprintf("expected FeatureGate object, but got object of type %T", old))
```
[the remark this answers is not in the document]
Comment (rajathagasthya) [reply apis/core/v1alpha2/readinessprovider_webhook.go:58]: Please use `field.*` errors like the feature gate webhook does.

https://github.com/vmware-tanzu/tanzu-framework/blob/main/apis/core/v1alpha2/featuregate_webhook.go#L128
https://github.com/vmware-tanzu/tanzu-framework/blob/main/apis/core/v1alpha2/featuregate_webhook.go#L252

File: apis/core/v1alpha2/readinessprovider_webhook.go [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+	}
+	if oldObj == nil {
+		return nil
+	}
+
```
Comment (rajathagasthya) [inline apis/core/v1alpha2/readinessprovider_webhook.go:63]: We still need to validate here all the fields we validate in create as well.

File: apis/core/v1alpha2/readinessprovider_webhook.go (hunk 1 above)
Comment (sathyanarays) [reply apis/core/v1alpha2/readinessprovider_webhook.go:40]: That's not necessary. The readiness providers can exist independently.

For example, we have many packages in TKG standard packages. Each of these packages can contain a readiness provider. But, some of the installed packages may not be a part of any `Readiness` definition that the cluster administrator has defined. I can have a grafana package installed along with its readiness provider; but, I may not have a readiness definition that is interested in knowing grafana's state.

We don't do any validation of check names or check refs in the readiness controller. The check names and the associated providers should be validated and approved at the org level before being deployed to the cluster by the cluster administrator. We can take an analogy from the Kubernetes cluster-role resource where we can provide any arbitrary API group and resource; the validation is not done to check if the mentioned API group and resource actually exist on the cluster.

File: readiness/controller/pkg/readinessprovider/suite_test.go [hunk 6]
```diff
@@ -38,13 +45,51 @@ var setupLog = ctrl.Log.WithName("controllers").WithName("readinessprovider")
...
+var generatedWebhookManifestBytes []byte
 
 func TestAPIs(t *testing.T) {
 	RegisterFailHandler(Fail)
 
 	RunSpecs(t, "Controller Suite")
 }
 
+func generateCertificateAndManifests() error {
+	key, cert, err := testutil.GenerateSelfSignedCertsForTest("tanzu-readinessprovider-webhook-service.default.svc")
```
[the remark this answers is not in the document]
Comment (sathyanarays) [reply readiness/controller/pkg/readinessprovider/suite_test.go]: Done.

File: apis/core/v1alpha2/readinessprovider_webhook.go (hunk 5 above)
Comment (sathyanarays) [reply apis/core/v1alpha2/readinessprovider_webhook.go]: Done

File: apis/core/v1alpha2/readinessprovider_webhook.go [hunk 7]
```diff
@@ -0,0 +1,71 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
```
[the remark this answers is not in the document]
Comment (sathyanarays) [reply apis/core/v1alpha2/readinessprovider_webhook.go]: Corrected

File: apis/core/v1alpha2/readinessprovider_webhook.go [hunk 8]
```diff
@@ -0,0 +1,54 @@
...
+// ValidateDelete implements webhook.Validator so a webhook will be registered for the type
+func (r *ReadinessProvider) ValidateDelete() error {
+	readinessproviderlog.Info("validate delete", "name", r.Name)
+	return nil
+}
+
+func (r *ReadinessProvider) validateObject() error {
+	for _, condition := range r.Spec.Conditions {
+		if condition.ResourceExistenceCondition == nil {
+			return apierrors.NewBadRequest(fmt.Sprintf("Expected condition %s to have exactly one type defined", condition.Name))
```
[the remark this answers is not in the document]
Comment (sathyanarays) [reply apis/core/v1alpha2/readinessprovider_webhook.go:49]: Will have a look and make changes

[the remark this answers is not in the document]
Comment (sathyanarays) [reply apis/core/v1alpha2/readinessprovider_webhook.go]: Done

File: readiness/controller/pkg/readinessprovider/suite_test.go [hunk 9]
```diff
@@ -114,6 +182,13 @@ var _ = AfterSuite(func() {
 })
 
 var _ = Describe("Readiness Provider controller", func() {
+	It("should fail when the condition type is empty", func() {
```
Comment (avi-08) [inline readiness/controller/pkg/readinessprovider/suite_test.go:185]: Can we add tests for Update and Delete as well.

File: readiness/controller/pkg/readinessprovider/suite_test.go [hunk 10]
```diff
@@ -114,6 +182,13 @@ var _ = AfterSuite(func() {
 })
 
 var _ = Describe("Readiness Provider controller", func() {
```
Comment (avi-08) [inline readiness/controller/pkg/readinessprovider/suite_test.go:184]: Can we add tests for Update and Delete as well.

File: packages/readiness/bundle/config/values.yaml [hunk 11]
```diff
@@ -5,3 +5,5 @@ deployment:
   hostNetwork: false
   nodeSelector: null
   tolerations: []
+  webhookServerPort: 9443
+  tlsCipherSuites: "TLS_ECDHE_ECDSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_ECDSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256,TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384"
```
Comment (avi-08) [inline packages/readiness/bundle/config/values.yaml:9]: Please add description for the new values in `packages/readiness/README.md`.

File: readiness/controller/go.mod [hunk 12]
```diff
@@ -5,6 +5,8 @@ go 1.19
 replace (
 	github.com/vmware-tanzu/tanzu-framework/apis/core => ../../apis/core
 	github.com/vmware-tanzu/tanzu-framework/capabilities/client => ../../capabilities/client
+	github.com/vmware-tanzu/tanzu-framework/featuregates/controller => ../../featuregates/controller
```
Comment (avi-08) [inline readiness/controller/go.mod:8]: Is this needed ?

File: apis/core/v1alpha2/readinessprovider_webhook.go (hunk 1 above)
Comment (avi-08) [reply apis/core/v1alpha2/readinessprovider_webhook.go:30]: This makes sense as ReadinessProvider could be a precursor to Readiness definition. I'm good with keeping checkRefs out of validation.

File: packages/readiness/bundle/config/values.yaml (hunk 11 above)
Comment (sathyanarays) [reply packages/readiness/bundle/config/values.yaml:9]: Done

File: readiness/controller/pkg/readinessprovider/suite_test.go (hunk 10 above)
Comment (sathyanarays) [reply readiness/controller/pkg/readinessprovider/suite_test.go:184]: Done

**Review by avi-08 — APPROVED** [review]
LGTM 👍

**Review by rajathagasthya — APPROVED** [review]

**Pull request merged**